libsolidity/analysis/PostTypeChecker.cpp

```diff
@@ -421,6 +434,111 @@ struct ReservedErrorSelector: public PostTypeChecker::Checker
+	void endVisit(Identifier const& _identifier) override
+	{
+		solAssert(_identifier.annotation().referencedDeclaration);
+		if (
+			*_identifier.annotation().referencedDeclaration == *m_declaration &&
+			_identifier.annotation().willBeWrittenTo
+		)
+			m_willBeWrittenTo = true;
+	}
+	void endVisit(InlineAssembly const& _inlineAssembly) override
+	{
+		YulLValueChecker yulChecker{m_declaration->name()};
+		yulChecker(_inlineAssembly.operations());
+		m_willBeWrittenTo = yulChecker.willBeWrittenTo();
+	}
+private:
+	Declaration const* m_declaration{};
+	bool m_willBeWrittenTo = false;
+};
+
+struct SimpleCounterForLoopChecker: public PostTypeChecker::Checker
+{
+	SimpleCounterForLoopChecker(ErrorReporter& _errorReporter): Checker(_errorReporter) {}
+	bool visit(ForStatement const& _forStatement) override
+	{
+		_forStatement.annotation().isSimpleCounterLoop = isSimpleCounterLoop(_forStatement);
+		return true;
+	}
+	bool isSimpleCounterLoop(ForStatement const& _forStatement) const
+	{
+		auto const* simpleCondition = dynamic_cast<BinaryOperation const*>(_forStatement.condition());
+		if (!simpleCondition || simpleCondition->getOperator() != Token::LessThan || simpleCondition->userDefinedFunctionType())
+			return false;
+		if (!_forStatement.loopExpression())
+			return false;
+
+		auto const* simplePostExpression = dynamic_cast<UnaryOperation const*>(&_forStatement.loopExpression()->expression());
+		// This matches both operators ++i and i++
+		if (!simplePostExpression || simplePostExpression->getOperator() != Token::Inc || simplePostExpression->userDefinedFunctionType())
+			return false;
+
+		auto const* lhsIdentifier = dynamic_cast<Identifier const*>(&simpleCondition->leftExpression());
+		auto const* lhsIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->leftExpression().annotation().type);
+		auto const* commonIntegerType = dynamic_cast<IntegerType const*>(simpleCondition->annotation().commonType);
+
+		if (!lhsIdentifier || !lhsIntegerType || !commonIntegerType || *lhsIntegerType != *commonIntegerType)
+			return false;
+
+		if (
+			!dynamic_cast<Identifier const*>(&simpleCondition->rightExpression()) &&
+			!dynamic_cast<Literal const*>(&simpleCondition->rightExpression())
+		)
+		{
+			solAssert(lhsIdentifier);
+			LValueChecker rhsConditionChecker{*lhsIdentifier};
+			simpleCondition->rightExpression().accept(rhsConditionChecker);
+			if (rhsConditionChecker.willBeWrittenTo())
+				return false;
+		}
+
+		auto const* incExpressionIdentifier = dynamic_cast<Identifier const*>(&simplePostExpression->subExpression());
+		if (
+			incExpressionIdentifier &&
+			incExpressionIdentifier->annotation().referencedDeclaration != lhsIdentifier->annotation().referencedDeclaration
+		)
+			return false;
+
+		if (
+			auto const* incVariableDeclaration = dynamic_cast<VariableDeclaration const*>(
+				incExpressionIdentifier->annotation().referencedDeclaration
+			);
+			!incVariableDeclaration->isLocalVariable()
```

Comment:
Shouldn't we check that `incVariableDeclaration` isn't `nullptr` before dereferencing?

Reply:
It is already checked. This is equivalent to 
```
auto const* incVariableDeclaration = ... ;
if (incVariableDeclaration)
{ ... }
```

Reply:
Ah, fair enough, I assumed that the assignment expression in the if statement would not be checked, but you are right.

Reply:
You are right, it is not the same when you have the semicolon.
Sorry about that, especially to @nikola-matic who first spotted it.

Reply:
> I don't think this is true. If you include the `;` then the declaration is not checked as a condition. At least that's what I always assumed.
> 
> I wasn't 100% sure myself since I rarely use that form, so I checked just in case with a small example and it seems to confirm that:
> 
> ```c++
> #include <iostream>
> 
> int main() {
>     bool b = true;
>     if (bool a = false; b)
>         std::cout << "true" << std::endl;
>     else
>         std::cout << "false" << std::endl;
> }
> ```
> 
> ```
> true
> ```
> 
> The result is "true" despite so the condition must be `b`, not `a && b`.

Ha, I actually did the same thing - the docs say that `init-statement` can be `nullptr`, but doesn't explicitly state how it affects the `if` as a whole, and didn't feel like digging through the standard, so ended up writing this example:

```c++
class C {
public:
	bool foo() { return true; }
};

int main()
{
	if (C *c = nullptr; c->foo()) {}
	return 0;
}
```
which did not in fact segfault, which is why I then assumed that Matheus was in fact right. So now that you posted your comment - I started extending the example a bit to see if I'm unable to write a simple program that will segfault. Figured maybe it's because the actual `if` block is empty - so the compiler will just optimize it away, since `c` is in local scope anyway, and `foo()` (even though not explicitly `const`, is still likely treated as `const`, since it's pretty easy for the compiler to figure this out). So I added a `cout` to to the block, recompiled, ran it again, and still no segfault.
Then, I added another `nullptr` dereference by adding
```
C *a = nullptr;
c-> foo();
```
before the `if` statement (so right in `main`), and still no segfault. Then, I figured, I have to add a member variable, and return *it* instead of just a `true` literal - and sure enough, this finally segfaulted. I tested it both with `gcc` and `clang`, and the behaviour is consistent between them, so even though dereferencing a `nullptr` is clearly undefined behaviour, the compiler implementers decided that it makes perfect sense to do the most unexpected thing when dereferencing a nullptr.

Thanks for coming to my TED talk, I sometime really hate this language.

Reply:
It's not like the standard isn't perfectly clear in that
```
if (init; cond) ...
```
is equivalent to
```
{
  init;
  if (cond) ...
}
```
:-)